contracts/base/BaseLiquidityManagement.sol

```diff
@@ -236,41 +203,40 @@ contract BaseLiquidityManagement is IBaseLiquidityManagement, SafeCallback {
         );
     }
 
+    /// Any outstanding amounts owed to the caller from the underlying modify call must be collected explicitly with `collect`.
     function _decreaseLiquidity(
         address owner,
         LiquidityRange memory range,
         uint256 liquidityToRemove,
         bytes memory hookData
-    ) internal returns (BalanceDelta delta) {
+    ) internal returns (BalanceDelta callerDelta, BalanceDelta thisDelta) {
         (BalanceDelta liquidityDelta, BalanceDelta totalFeesAccrued) =
             _modifyLiquidity(owner, range, -(liquidityToRemove.toInt256()), hookData);
 
-        // take all tokens first
-        // do NOT take tokens directly to the owner because this contract might be holding fees
-        // that need to be paid out (position.tokensOwed)
-        if (liquidityDelta.amount0() > 0) {
-            range.poolKey.currency0.take(manager, address(this), uint128(liquidityDelta.amount0()), true);
-        }
-        if (liquidityDelta.amount1() > 0) {
-            range.poolKey.currency1.take(manager, address(this), uint128(liquidityDelta.amount1()), true);
-        }
+        Position storage position = positions[owner][range.toId()];
 
-        // when decreasing liquidity, the user collects: 1) principal liquidity, 2) new fees, 3) old fees (position.tokensOwed)
+        // Calculates the fee growth since the last time the positions feeGrowthInside was updated
+        // Also updates the position's the feeGrowthInsideLast variables in storage.
+        (BalanceDelta callerFeesAccrued) = _updateFeeGrowth(range, position);
 
-        Position storage position = positions[owner][range.toId()];
-        BalanceDelta callerFeesAccrued = _updateFeeGrowth(range, position);
-        BalanceDelta principalDelta = liquidityDelta - totalFeesAccrued;
+        // Account for fees accrued to other users on the same range.
+        (callerDelta, thisDelta) = liquidityDelta.split(callerFeesAccrued, totalFeesAccrued);
 
-        // new fees = new fees + old fees + principal liquidity
-        callerFeesAccrued = callerFeesAccrued
-            + toBalanceDelta(uint256(position.tokensOwed0).toInt128(), uint256(position.tokensOwed1).toInt128())
-            + principalDelta;
+        BalanceDelta tokensOwed;
 
-        position.tokensOwed0 = 0;
-        position.tokensOwed1 = 0;
-        position.liquidity -= liquidityToRemove;
+        // Flush the callerDelta, incrementing the tokensOwed to the user and the amount claimable to this contract.
+        if (callerDelta.amount0() > 0) {
+            (tokensOwed, callerDelta, thisDelta) =
+                _moveCallerDeltaToTokensOwed(true, tokensOwed, callerDelta, thisDelta);
+        }
 
-        return callerFeesAccrued;
+        if (callerDelta.amount1() > 0) {
+            (tokensOwed, callerDelta, thisDelta) =
+                _moveCallerDeltaToTokensOwed(false, tokensOwed, callerDelta, thisDelta);
+        }
```

Comment:
certain that the `callerDelta.amount1() > 0` isnt needed since decrease liquidity should always create a credit / positive delta

---

noting for myself and can fix with transient library

Reply:
Nice yeah, I think I put that there in case it ever was exactly 0 to just skip all that..

since technically someone could call decrease with liquidityAmount = 0?